pkg/observable/channel/observable.go

```diff
@@ -0,0 +1,148 @@
+package channel
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"pocket/pkg/observable"
+)
+
+var _ observable.Observable[any] = &channelObservable[any]{}
+
+// option is a function which receives and can modify the channelObservable state.
+type option[V any] func(obs *channelObservable[V])
+
+// channelObservable implements the observable.Observable interface and can be notified
+// via its corresponding producer channel.
+type channelObservable[V any] struct {
+	// producer is an observable-wide channel that is used to receive values
+	// which are subsequently re-sent to observers.
+	producer chan V
+	// observersMu protects observers from concurrent access/updates
+	observersMu *sync.RWMutex
+	// observers is a list of channelObservers that will be notified when producer
+	// receives a value. It is a pointer because ...
+	observers []*channelObserver[V]
+}
+
+// NewObservable creates a new observable is notified when the producer channel
+// receives a value.
+// func NewObservable[V any](producer chan V) (observable.Observable[V], chan<- V) {
+func NewObservable[V any](opts ...option[V]) (observable.Observable[V], chan<- V) {
+	// initialize an observer that publishes messages from 1 producer to N observers
+	obs := &channelObservable[V]{
+		observersMu: &sync.RWMutex{},
+		observers:   []*channelObserver[V]{},
+	}
+
+	for _, opt := range opts {
+		opt(obs)
+	}
+
+	// if the caller does not provide a producer, create a new one and return it
+	if obs.producer == nil {
+		obs.producer = make(chan V)
+	}
+
+	// start listening to the producer and emit values to observers
+	go obs.goListen(obs.producer)
+
+	return obs, obs.producer
+}
+
+// WithProducer returns an option function sets the given producer in an observable
+// when passed to NewObservable().
+func WithProducer[V any](producer chan V) option[V] {
+	return func(obs *channelObservable[V]) {
+		obs.producer = producer
+	}
+}
+
+// Subscribe returns an observer which is notified when the producer channel
+// receives a value.
+func (obsvbl *channelObservable[V]) Subscribe(ctx context.Context) observable.Observer[V] {
+	obsvbl.observersMu.Lock()
+	defer obsvbl.observersMu.Unlock()
+
+	observer := NewObserver[V](ctx, obsvbl.onUnsubscribeFactory)
+	obsvbl.observers = append(obsvbl.observers, observer)
+
+	// caller can rely on context cancellation or call Close() to unsubscribe
+	// active observers
+	if ctx != nil {
+		// asynchronously wait for the context to close and unsubscribe
+		go goUnsubscribeOnDone[V](ctx, observer)
+	}
```

Comment:
This can be moved into `NewObservable()`. Since `channelObservable` has a reference to its `observers`, we can create a single go routine in a higher scope and iterate over the observers to unsubscribe them all at once.

Reply:
I decided against this after all because a consequence of doing so would be either coupling all observers to the same context, which seems wrong, or  having an optional context in both  the constructor function and  the `#Subscribe()` method, which seems unnecessarily complex IMHO.